libmultipath/public_api.c

```diff
@@ -0,0 +1,794 @@
+	if (pgp->priority < 0)
+		mp_pg->priority = 0;
+	else
+		mp_pg->priority = pgp->priority & UINT32_MAX;
+	if (pgp->selector != NULL) {
+		mp_pg->selector = STRDUP(pgp->selector);
+		if (mp_pg->selector == NULL)
+			return 1;
+	}
+	return 0;
+}
+
+static int _fill_dmmp_path(struct dmmp_path *mp_path, struct path *pp)
+{
+	strcpy(mp_path->name, pp->dev);
+	if (pp->state > 0)
+		mp_path->status = pp->state;
+	return 0;
+}
+
+static int _update_path_status(struct multipath *mpp)
+{
+	int i, j;
+	struct pathgroup * pgp;
+	struct path * pp;
+
+	if (!mpp->pg)
+		return 0;
+
+	vector_foreach_slot (mpp->pg, pgp, i) {
+		if (!pgp->paths)
+			continue;
+
+		vector_foreach_slot (pgp->paths, pp, j) {
+			if (!strlen(pp->dev)) {
+				if (devt2devname(pp->dev, FILE_NAME_SIZE,
+						 pp->dev_t)) {
+					/*
+					 * path is not in sysfs anymore
+					 */
+					pp->chkrstate = pp->state = PATH_DOWN;
+					continue;
+				}
+				pp->mpp = mpp;
+				if (pathinfo(pp, conf->hwtable, DI_ALL))
+					pp->state = PATH_UNCHECKED;
+				continue;
+			}
+			pp->mpp = mpp;
+			if (pp->state == PATH_UNCHECKED ||
+			    pp->state == PATH_WILD) {
+				if (pathinfo(pp, conf->hwtable, DI_CHECKER))
+					pp->state = PATH_UNCHECKED;
+			}
+
+			if (pp->priority == PRIO_UNDEF) {
+				if (pathinfo(pp, conf->hwtable, DI_PRIO))
+					pp->priority = PRIO_UNDEF;
+			}
+		}
+	}
+	return 0;
+}
+
+static void _dmmp_path_group_list_free(struct dmmp_path_group **mp_pgs,
+				      uint32_t mp_pg_count)
+{
+	uint32_t i = 0;
+	struct dmmp_path_group *mp_pg = NULL;
+
+	if (mp_pgs == NULL)
+		return;
+	for (i = 0; i < mp_pg_count; ++i) {
+		mp_pg = mp_pgs[i];
+		_dmmp_path_group_free(mp_pg);
+	}
+	FREE(mp_pgs);
+}
+
+static void _dmmp_path_group_free(struct dmmp_path_group *mp_pg)
+{
+	uint32_t i = 0;
+	if (mp_pg == NULL)
+		return;
+	if (mp_pg->selector != NULL)
+		FREE((char *) mp_pg->selector);
+
+	if (mp_pg->mp_paths != NULL) {
+		for (i =0; i < mp_pg->mp_path_count; ++i) {
+			_dmmp_path_free(mp_pg->mp_paths[i]);
+		}
+		FREE(mp_pg->mp_paths);
+	}
+	FREE(mp_pg);
+}
+
+static void _dmmp_path_free(struct dmmp_path *mp_path)
+{
+	if (mp_path == NULL)
+		return;
+	FREE(mp_path);
+}
+
+static enum devtypes _get_dev_type(char *dev)
+{
+	struct stat buf;
+	int i;
+
+	if (stat(dev, &buf) == 0 && S_ISBLK(buf.st_mode)) {
+		if (dm_is_dm_major(major(buf.st_rdev)))
+			return DEV_DEVMAP;
+		return DEV_DEVNODE;
+	}
+	else if (sscanf(dev, "%d:%d", &i, &i) == 2)
+		return DEV_DEVT;
+	else
+		return DEV_DEVMAP;
+}
+
+
+int dmmp_mpath_list(struct dmmp_mpath ***mpaths, uint32_t *mpath_count)
+{
+	struct multipath *mpp = NULL;
+	struct dmmp_mpath *mpath = NULL;
+	vector mppvec = NULL;
+	vector pathvec = NULL;
+	int rc = 0;
+	int i = 0;
+
+	*mpaths = NULL;
+	*mpath_count = 0;
+
+	if (_init() != 0)
+		goto error_out;
+
+	mppvec = vector_alloc();
+
+	rc = dm_get_maps (mppvec);
+	if (rc != 0)
+		goto error_out;
+
+	*mpaths = (struct dmmp_mpath **)
+		MALLOC(sizeof(struct dmmp_mpath *) * VECTOR_SIZE(mppvec));
```

Comment:
The `mpath->mp_pg_count` and `mpath->mp_pgs` are holding the path groups.
The paths are saved in `struct dmmp_path_group` as:
`mp_pg->mp_paths` and `mp_pg->mp_path_count`.